core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -599,21 +606,30 @@ private[log] class Cleaner(val id: Int,
         }
         currentSegmentOpt = nextSegmentOpt
       }
-
-      cleaned.onBecomeInactiveSegment()
-      // flush new segment to disk before swap
-      cleaned.flush()
-
-      // update the modification date to retain the last modified date of the original files
-      val modified = segments.last.lastModified
-      cleaned.lastModified = modified
-
-      // swap in new segment
-      info(s"Swapping in cleaned segment $cleaned for segment(s) $segments in log $log")
-      log.replaceSegments(List(cleaned), segments)
+      
+      // Result of cleaning included at least one record.
+      if (cleanedSegment.isDefined) {
+        val cleaned = cleanedSegment.get
+        cleaned.onBecomeInactiveSegment()
+        // flush new segment to disk before swap
+        cleaned.flush()
+
+        // update the modification date to retain the last modified date of the original files
+        val modified = segments.last.lastModified
+        cleaned.lastModified = modified
+
+        // swap in new segment
+        info(s"Swapping in cleaned segment $cleaned for segment(s) $segments in log $log")
+        log.replaceSegments(List(cleaned), segments)
```

Comment:
Would it make more sense to update log start offset in `replaceSegments` when we are protected by the lock? Otherwise, it seems like the log start offset will be out of sync with the segments until all of them have been cleaned and `maybeIncrementLogStartOffset` is called. 

Note that `deleteSegments` does update the log start offset. Interestingly, `SegmentDeletion` is used for the reason in this case, and not `SegmentCompaction`.

Reply:
Right. I wasn't sure if it made sense to change the deleteSegments code to support the SegmentCompaction reason. The entire segment is being deleted here, but I see how this may be confusing when reading in the log.

Reply:
As for the first part about `replaceSegments`. It seems that there are a few other times we call this method. I'm wondering if we would want to update the logStartOffset in these cases (probably) and what the reason should be. I'm also wondering if we should include the reason as a parameter to `replaceSegments`.

Reply:
There are really only two cases for `replaceSegments`: log compaction, and segment splitting. We also call it in `completeSwapOperations`, but that is for recovery if the broker crashes in the middle of `replaceSegments`. It should be safe in either case to adjust the log start offset. Probably something similar to the deletion case which just ensures that the log start offset did not fall off the log:
```scala
maybeIncrementLogStartOffset(segments.firstEntry.getValue.baseOffset, SegmentDeletion)
```
Ideally `LogCompaction` would be the reason, at least in the log compaction case. It's becoming cumbersome to pass through the reason in all cases it is used, but I'm not sure I see an alternative. Perhaps we can consider how to refactor this in the future. Maybe there should be some kind of context object which encapsulates the higher level operation that is being done (such as log compaction or retention enforcement).